pkg/controller/externalippool/controller.go

```diff
@@ -0,0 +1,417 @@
+// Copyright 2021 Antrea Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package externalippool
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/retry"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+
+	antreacrds "antrea.io/antrea/pkg/apis/crd/v1alpha2"
+	clientset "antrea.io/antrea/pkg/client/clientset/versioned"
+	antreainformers "antrea.io/antrea/pkg/client/informers/externalversions/crd/v1alpha2"
+	antrealisters "antrea.io/antrea/pkg/client/listers/crd/v1alpha2"
+	"antrea.io/antrea/pkg/controller/externalippool/ipallocator"
+	"antrea.io/antrea/pkg/controller/metrics"
+)
+
+const (
+	controllerName = "ExternalIPPoolController"
+	// Set resyncPeriod to 0 to disable resyncing.
+	resyncPeriod time.Duration = 0
+	// How long to wait before retrying the processing of an ExternalIPPool change.
+	minRetryDelay = 5 * time.Second
+	maxRetryDelay = 300 * time.Second
+	// Default number of workers processing an ExternalIPPool change.
+	defaultWorkers = 4
+)
+
+var (
+	ErrExternalIPPoolNotFound = errors.New("ExternalIPPool not found")
+)
+
+type ExternalIPAllocator interface {
+	// AddConsumer adds a consumer for ExternalIPPool events. It will block other consumers from allocating new IPs by
+	// AllocateIPFromPool() or AllocateIP() until it calls ConsumerRestoreFinished().
+	AddConsumer(func(string))
+	// ConsumerRestoreFinished is used to indicates that the consumer has finished updating its allocated IPs.
+	ConsumerRestoreFinished()
```

Comment:
I like the idea of using a single call `RestoreIPAllocations` to make the restoration process more precise. But seems that it will make it harder to distinguish the succeeded and the failed IP allocations restoration. Do you have more insights?

Reply:
You could return the failed allocations